### PR TITLE
[BLUE] Optimize ScrIoControl for speed

### PR DESCRIPTION
**Pull request by katahiromz**
## Purpose

JIRA issue: [CORE-18838](https://jira.reactos.org/browse/CORE-18838)

## Proposed changes

- Optimize `IOCTL_CONSOLE_FILL_OUTPUT_ATTRIBUTE` and `IOCTL_CONSOLE_FILL_OUTPUT_CHARACTER` for speed by using cache.

## TODO

- [x] Do tests.

### CODE REVIEW

**Comment by HBelusca** [comment]
Please give speed measurements. Otherwise the PR will be rejected.

**Comment by katahiromz** [comment]
```txt
BEFORE (#undef OPTIMIZED):
(base/setup/usetup/usetup.c:4471) CONSOLE_ClearScreen x 100: 33589248

AFTER: (#define OPTIMIZED):
(base/setup/usetup/usetup.c:4471) CONSOLE_ClearScreen x 100: 22042944

(33589248-22042944)/33589248 == 34.4%
```

**Comment by Extravert-ir** [comment]
What did you change? What it "direct pointer access"?

**Comment by katahiromz** [comment]
```txt
BEFORE:
2nd Trial: (base/setup/usetup/usetup.c:4471) CONSOLE_ClearScreen x 100: 31489920
3rd Trial: (base/setup/usetup/usetup.c:4471) CONSOLE_ClearScreen x 100: 28640832
4th Trial: (base/setup/usetup/usetup.c:4471) CONSOLE_ClearScreen x 100: 29990400

AFTER:
2nd Trial: (base/setup/usetup/usetup.c:4471) CONSOLE_ClearScreen x 100: 22492800
3rd Trial: (base/setup/usetup/usetup.c:4471) CONSOLE_ClearScreen x 100: 20543424
4th Trial: (base/setup/usetup/usetup.c:4471) CONSOLE_ClearScreen x 100: 22642752

Comparison:
2nd Trial: (31489920-22492800)/31489920 == 28.6%
3rd Trial: (28640832-20543424)/28640832 == 28.3%
4th Trial: (29990400-22642752)/29990400 == 24.5%
```

**Comment by katahiromz** [comment]
@Extravert-ir See this:
https://github.com/reactos/reactos/pull/5185/commits/fdb4b9fe28d4a112e6e67b80b43b306937face8a

**Comment by Extravert-ir** [comment]
> @Extravert-ir See this: https://github.com/reactos/reactos/commit/fdb4b9fe28d4a112e6e67b80b43b306937face8a

Why did you copy the whole function, instead of keeping that change as a PR? :)

**Comment by katahiromz** [comment]
@Extravert-ir Fixed.

File: drivers/setup/blue/blue.c [hunk 1]
```diff
@@ -1073,16 +1075,28 @@ ScrIoControl(
 
             if (DeviceExtension->Enabled && DeviceExtension->VideoMemory)
             {
+#ifdef OPTIMIZED
+                CHAR attr = Buf->wAttribute;
```
Comment (Extravert-ir) [inline drivers/setup/blue/blue.c:1079]: The real optimization is only here (caching global variable as a local one) and at other places like that.
Btw, did you test an optimized build? The difference will probably be way smaller

Comment (katahiromz) [reply drivers/setup/blue/blue.c:1079]: > did you test an optimized build?

Yep, I did. I feel quicker.

Comment (HBelusca) [reply drivers/setup/blue/blue.c:1079]: And what about an optimized build without your changes?

Comment (katahiromz) [reply drivers/setup/blue/blue.c:1079]: The Release build is not tried yet.

Comment (katahiromz) [reply drivers/setup/blue/blue.c:1079]: image
`error D8016` happened.

Comment (HBelusca) [reply drivers/setup/blue/blue.c:1079]: Yes, add `-DRUNTIME_CHECKS=OFF` to the configure command line too.

Comment (katahiromz) [reply drivers/setup/blue/blue.c:1079]: image

Comment (katahiromz) [reply drivers/setup/blue/blue.c:1079]: image
Fxxk again.

Comment (katahiromz) [reply drivers/setup/blue/blue.c:1079]: image
Okay, okay.

Comment (katahiromz) [reply drivers/setup/blue/blue.c:1079]:
```txt
In optimized build:

BEFORE:
(base/setup/usetup/usetup.c:4471) CONSOLE_ClearScreen x 100: 39437376
(base/setup/usetup/usetup.c:4471) CONSOLE_ClearScreen x 100: 33289344
(base/setup/usetup/usetup.c:4471) CONSOLE_ClearScreen x 100: 34788864
(base/setup/usetup/usetup.c:4471) CONSOLE_ClearScreen x 100: 29840448

AFTER:
(base/setup/usetup/usetup.c:4471) CONSOLE_ClearScreen x 100: 27741120
(base/setup/usetup/usetup.c:4471) CONSOLE_ClearScreen x 100: 26541504
(base/setup/usetup/usetup.c:4471) CONSOLE_ClearScreen x 100: 29840448
(base/setup/usetup/usetup.c:4471) CONSOLE_ClearScreen x 100: 25491840

Comparison:
(39437376-27741120)/39437376 == 29.7%.
(33289344-26541504)/33289344 == 20.3%.
(34788864-29840448)/34788864 == 14.2%.
(29840448-25491840)/29840448 == 14.6%.
```

Comment (HBelusca) [inline drivers/setup/blue/blue.c:1079]: When you cache wAttr, use `UCHAR` variable instead.

Comment (katahiromz) [reply drivers/setup/blue/blue.c]: Fixed.

File: drivers/setup/blue/blue.c [hunk 2]
```diff
@@ -1073,16 +1075,28 @@ ScrIoControl(
 
             if (DeviceExtension->Enabled && DeviceExtension->VideoMemory)
             {
+#ifdef OPTIMIZED
+                UCHAR attr = Buf->wAttribute;
+                PUCHAR pch;
```
Comment (HBelusca) [inline drivers/setup/blue/blue.c:1080]:
```suggestion
```

File: drivers/setup/blue/blue.c [hunk 3]
```diff
@@ -1073,16 +1075,28 @@ ScrIoControl(
...
                 vidmem = DeviceExtension->VideoMemory;
                 offset = (Buf->dwCoord.X + Buf->dwCoord.Y * DeviceExtension->Columns) * 2 + 1;
 
                 nMaxLength = min(nMaxLength,
                                  (DeviceExtension->Rows - Buf->dwCoord.Y)
                                     * DeviceExtension->Columns - Buf->dwCoord.X);
 
+#ifdef OPTIMIZED
+                pch = vidmem + offset;
+#endif
```
Comment (HBelusca) [inline drivers/setup/blue/blue.c:1091]:
```suggestion
```

File: drivers/setup/blue/blue.c [hunk 4]
```diff
@@ -1073,16 +1075,28 @@ ScrIoControl(
...
                 for (dwCount = 0; dwCount < nMaxLength; dwCount++)
                 {
+#ifdef OPTIMIZED
+                    *pch = attr;
+                    pch += 2;
```
Comment (HBelusca) [inline drivers/setup/blue/blue.c:1096]:
```suggestion
                    vidmem[offset + (dwCount * 2)] = attr;
```

File: drivers/setup/blue/blue.c [hunk 5]
```diff
@@ -1203,16 +1217,27 @@ ScrIoControl(
 
             if (DeviceExtension->Enabled && DeviceExtension->VideoMemory)
             {
+#ifdef OPTIMIZED
+                PUCHAR pch;
+#endif
```
Comment (HBelusca) [inline drivers/setup/blue/blue.c:1222]:
```suggestion
```

File: drivers/setup/blue/blue.c [hunk 6]
```diff
@@ -1203,16 +1217,27 @@ ScrIoControl(
...
                 vidmem = DeviceExtension->VideoMemory;
                 offset = (dwCoord.X + dwCoord.Y * DeviceExtension->Columns) * 2 + 1;
 
                 nMaxLength = min(nMaxLength,
                                  (DeviceExtension->Rows - dwCoord.Y)
                                     * DeviceExtension->Columns - dwCoord.X);
 
+#ifdef OPTIMIZED
+                pch = vidmem + offset;
+#endif
```
Comment (HBelusca) [inline drivers/setup/blue/blue.c:1232]:
```suggestion
```

File: drivers/setup/blue/blue.c [hunk 7]
```diff
@@ -1203,16 +1217,27 @@ ScrIoControl(
...
                 for (dwCount = 0; dwCount < nMaxLength; dwCount++, pAttr++)
                 {
+#ifdef OPTIMIZED
+                    *pch = *((PCHAR)pAttr);
+                    pch += 2;
+#else
```
Comment (HBelusca) [inline drivers/setup/blue/blue.c:1238]:
```suggestion
```

File: drivers/setup/blue/blue.c [hunk 8]
```diff
@@ -1203,16 +1217,27 @@ ScrIoControl(
...
                     vidmem[offset + (dwCount * 2)] = *((PCHAR)pAttr);
+#endif
```
Comment (HBelusca) [inline drivers/setup/blue/blue.c:1240]:
```suggestion
```

File: drivers/setup/blue/blue.c [hunk 9]
```diff
@@ -1271,16 +1296,28 @@ ScrIoControl(
 
             if (DeviceExtension->Enabled && DeviceExtension->VideoMemory)
             {
+#ifdef OPTIMIZED
+                CHAR ch = Buf->cCharacter;
+                PUCHAR pch;
```
Comment (HBelusca) [inline drivers/setup/blue/blue.c:1301]:
```suggestion
                UCHAR ch = Buf->cCharacter;
```

File: drivers/setup/blue/blue.c [hunk 10]
```diff
@@ -1271,16 +1296,28 @@ ScrIoControl(
...
                 vidmem = DeviceExtension->VideoMemory;
                 offset = (Buf->dwCoord.X + Buf->dwCoord.Y * DeviceExtension->Columns) * 2;
 
                 nMaxLength = min(nMaxLength,
                                  (DeviceExtension->Rows - Buf->dwCoord.Y)
                                     * DeviceExtension->Columns - Buf->dwCoord.X);
 
+#ifdef OPTIMIZED
+                pch = vidmem + offset;
+#endif
```
Comment (HBelusca) [inline drivers/setup/blue/blue.c:1312]:
```suggestion
```

File: drivers/setup/blue/blue.c [hunk 11]
```diff
@@ -1271,16 +1296,28 @@ ScrIoControl(
...
                 for (dwCount = 0; dwCount < nMaxLength; dwCount++)
                 {
+#ifdef OPTIMIZED
+                    *pch = ch;
+                    pch += 2;
```
Comment (HBelusca) [inline drivers/setup/blue/blue.c:1317]:
```suggestion
                    vidmem[offset + (dwCount * 2)] = ch;
```

**Review by HBelusca — CHANGES_REQUESTED** [review]
@katahiromz Can you please retest your PR with timing measurements, but with the following suggestions:
only the local capture of either `Buf->wAttribute` or `Buf->cCharacter` is kept, while the rest (`videomem[offset + ...]`) statys as it currently was.

**Comment by katahiromz** [comment]
> Can you please retest

```txt
BEFORE:
1st Trial (base/setup/usetup/usetup.c:4471) CONSOLE_ClearScreen x 100: 33589248
2nd Trial: (base/setup/usetup/usetup.c:4471) CONSOLE_ClearScreen x 100: 31489920
3rd Trial: (base/setup/usetup/usetup.c:4471) CONSOLE_ClearScreen x 100: 28640832
4th Trial: (base/setup/usetup/usetup.c:4471) CONSOLE_ClearScreen x 100: 29990400

AFTER:
1st Trial: (base/setup/usetup/usetup.c:4471) CONSOLE_ClearScreen x 100: 20393472
2nd Trial: (base/setup/usetup/usetup.c:4471) CONSOLE_ClearScreen x 100: 19343808
3rd Trial: (base/setup/usetup/usetup.c:4471) CONSOLE_ClearScreen x 100: 20693376
4th Trial: (base/setup/usetup/usetup.c:4471) CONSOLE_ClearScreen x 100: 18294144

Comparison:
1st: (33589248-20393472)/33589248 == 39.3%.
2nd: (31489920-19343808)/31489920 == 38.6%.
3rd: (28640832-20693376)/28640832 == 27.7%.
4th: (29990400-18294144)/29990400==39.0%.

Average: (39.3+38.6+27.7+39.0)/4 == 36.2%.
```

**Comment by HBelusca** [comment]
So indeed, @Extravert-ir was right and the performance gain only comes from the caching of the `Buf->wAttribute` and `Buf->cCharacter` :)

**Review by HBelusca — APPROVED** [review]
I think you can commit that -- Once you remove the temporary measurement code and keep only the OPTIMIZED parts.

**Pull request merged**